### PR TITLE
[FLINK-30623][runtime][bug] The canEmitBatchOfRecords should check recordWriter and changelogWriterAvailabilityProvider are available

### PR DESCRIPTION
**Pull request by 1996fanrui**
## What is the purpose of the change

[FLINK-30533](https://issues.apache.org/jira/browse/FLINK-30533) doesn't check `recordWriter.isAvailable()` in the loop of SourceOperator.

## Brief change log

  - 1st commit: [FLINK-30623][runtime][refactor] Using the `CanEmitBatchOfRecordsChecker` instead of `Supplier<Boolean>`
  - 2nd commit: [FLINK-30623][runtime][refactor] Using the `canEmitBatchOfRecords` instead of `isEmitNextLoopDisabled` in the SourceOperator
  - 3rd commit: [FLINK-30623][runtime][bug] The `canEmitBatchOfRecords` should check recordWriter and changelogWriterAvailabilityProvider are available

I ran the checkpoint benchmark in my local, and here is the performance report, I think the performance is recovered after this PR.

```
Before 2nd commit:

Benchmark                                              (mode)   Mode  Cnt   Score   Error  Units
CheckpointingTimeBenchmark.checkpointSingleInput      ALIGNED  thrpt   30   1.247 ± 0.177  ops/s
CheckpointingTimeBenchmark.checkpointSingleInput    UNALIGNED  thrpt   30  65.063 ± 9.090  ops/s
CheckpointingTimeBenchmark.checkpointSingleInput  UNALIGNED_1  thrpt   30  35.096 ± 4.313  ops/s

After 2nd commit:

Benchmark                                              (mode)   Mode  Cnt    Score    Error  Units
CheckpointingTimeBenchmark.checkpointSingleInput      ALIGNED  thrpt   30    1.345 ±  0.111  ops/s
CheckpointingTimeBenchmark.checkpointSingleInput    UNALIGNED  thrpt   30  299.665 ± 36.048  ops/s
CheckpointingTimeBenchmark.checkpointSingleInput  UNALIGNED_1  thrpt   30  183.998 ± 24.652  ops/s
```

## Verifying this change

- OneInputStreamTaskTest#testCanEmitBatchOfRecords
- TwoInputStreamTaskTest#testCanEmitBatchOfRecords
- MultipleInputStreamTaskTest#testCanEmitBatchOfRecords

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not documented

### CODE REVIEW

**Comment by 1996fanrui** [comment]
Hi @pnowojski @lindong28 @fredia , I have updated the PR to fix the bug that [FLINK-30533](https://issues.apache.org/jira/browse/FLINK-30533) doesn't check recordWriter.isAvailable() in the loop of SourceOperator. Please help take a look in your free time, thanks a lot ~

And I refactored the `canEmitBatchOfRecords` in the first commit according to @pnowojski 's [comment](https://issues.apache.org/jira/browse/FLINK-30533?focusedCommentId=17676585&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17676585). And after the second commit, the performance is recovered in my Mac.

**Comment by lindong28** [comment]
Thanks @1996fanrui for fixing the bug! The bug fix LGTM.

I am not sure if we should add the `DataOutput#canEmitBatchOfRecords` thought. Prior to this PR, `DataOutput` is responsible for data flow and it only provides API to emit data/watermark/latency. Classes such as `SourceOperator` and `StreamTask` are responsible for control flow (e.g. mailbox, availability, loop). Having `StreamTask` passes `mailboxHasMail` to `SourceOperator` preserves this separation of data flow and control flow.

This PR changes this abstraction by adding `DataOutput#canEmitBatchOfRecords`. Now `DataOutput` is responsible for control flow as well.

It seems cleaner and simpler to still have DataOutput be responsible only for data flow. Can you explain what is the benefit for this refactor?

**Comment by 1996fanrui** [comment]
Hi @lindong28 , thanks a lot for your review and discuss here.

> Could you explain what is the benefit for this refactor?

As I understand, there are two reasons for adding `canEmitBatchOfRecords` to `DataOutput`:
- `DataOutput` is responsible for emitting data/watermark/latency, and `canEmitBatchOfRecords` let caller know whether the `emit` can be called repeatedly, so it's related to the `emit`.
- The `canEmitBatchOfRecords` may be called in any emit place, such as: `SourceOperator` and `NetworkInput`, if we add `canEmitBatchOfRecords` in `DataOutput`, it's very helpful for code reuse. You can see the last commit of #21697 , I have made an improvement of `NetworkInput`. The `DataOutput#canEmitBatchOfRecords` is easy to reuse.

Please correct me if I'm wrong, thanks~ 

cc @pnowojski

**Comment by lindong28** [comment]
Hi @1996fanrui, thanks for the explanation.

Regarding the 1st point mentioned above, yes `canEmitBatchOfRecords()` is related to emit int the sense that it mangaes the "control flow" of when emit can be called. So it is probably consistent with my previous comment.

Regarding the 2nd point, I guess you are saying the implementation of this "optimization" after the refactor is simpler. IMO the abstraction (or separate of concern) between classes is generally more important than implementation, unless one implementation is considerably much simpler and easier to read (which is probably not the case here).

I guess ultimately it comes down to whether the extra implementation simplicity is more important than keeping the DataOutput focusing only on data flow. I am inclined to keep the existing abstraction. But I am open for the approach in this PR if both you and @pnowojski prefer this.

**Comment by 1996fanrui** [comment]
Hi @lindong28 , thanks for your quick feedback.

I want to check with you the first question: No matter which class canEmitBatchOfRecords is added to, we should use `Supplier<Boolean> canEmitBatchOfRecords` instead of `Supplier<Boolean> mailboxHasMail;` due to `canEmitBatchOfRecords` should add some other checks and it's more general and abstract, right?

If yes, we can focus on the next question: Where is `canEmitBatchOfRecords` added to?

> Regarding the 1st point mentioned above, yes `canEmitBatchOfRecords()` is related to emit int the sense that it mangaes the "control flow" of when emit can be called. So it is probably consistent with my previous comment.
> 
> Regarding the 2nd point, I guess you are saying the implementation of this "optimization" after the refactor is simpler. IMO the abstraction (or separate of concern) between classes is generally more important than implementation, because good abstraction makes code more intuitive and easier to maintain in the long term. I suppose the difference between the complexity of these two implementations is not significant in this case.
> 
> I guess ultimately it comes down to whether the extra implementation simplicity is more important than keeping the DataOutput focusing only on data flow. I am inclined to keep the existing abstraction. 

Sounds make sense. Let's wait for @pnowojski 's feedback.

**Comment by lindong28** [comment]
Hi @1996fanrui, yes I agree we should use Supplier<Boolean> canEmitBatchOfRecords.

I believe `canEmitBatchOfRecords` can be passed to `SourceOperator` similar to how `mailboxHasMail ` is passed to `SourceOperator`. Can you see if the approach in this commit [1] works?

[1] https://github.com/apache/flink/compare/master...lindong28:flink:use-task-is-available

**Comment by 1996fanrui** [comment]
> Hi @1996fanrui, yes I agree we should use Supplier canEmitBatchOfRecords.
> 
> I believe `canEmitBatchOfRecords` can be passed to `SourceOperator` similar to how `mailboxHasMail ` is passed to `SourceOperator`. Can you see if the approach in this commit [1] works?
> 
> [1] [lindong28@bbd2055](https://github.com/lindong28/flink/commit/bbd2055f7d6a203af69eeb1879718f953b2bfffd)

I created a POC branch[1] , please see the latest two commits. The first commit is similar to your commit. And I found the `isEmitNextLoopDisabled` can be simplify, so I using the `canEmitBatchOfRecords` instead of `isEmitNextLoopDisabled` in the 2nd commit. WDYT?

[1] https://github.com/1996fanrui/flink/commits/30623/canEmitBatchOfRecords-POC

**Comment by lindong28** [comment]
Thanks @1996fanrui for the POC. The approach in the 2nd commit looks much simpler than using the `isEmitNextLoopDisabled()`.

**Comment by pnowojski** [comment]
I think it's fine for the `DataOutput` to instruct the caller if it can be called again. Actually maybe even instead of adding `DataOutput#canEmitBatchOfRecords`, maybe `DataOutput#emitRecord` should return a `boolean` (most likely no need to change the watermark/latency marker versions). 

About the `Suppliers`, I don't like using them personally tbh. For one thing they are hard to follow in the IDE (finding implementations or usages do not work). This could be mitigated with using a named interface instead of `Supplier<Boolean>`.  Also it may suggest that there is some weird spaghetti dependency in the code, where some components are somehow sub-optimally designed. @1996fanrui's proposal is still using a supplier, but less so compared to master. 

The other thing that looks strange if I take a step back, is that we now have to different `while (true)` loops. One in the `MailboxProcessor#runMailboxLoop` the other in `SourceOperator` and with FLINK-30709 in the network source as well, doing basically the same thing. I will think about this a bit more.

**Comment by lindong28** [comment]
@pnowojski In order to address the problem around following this supplier in IDE, how about we add an internal functional interface in `StreamTask`? It could be similar to the existing `SizeSupplier` interface ([link](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java#L294)).

Suppose the issue around the IDE can be addressed like this, is there any other concern with the POC approach described here [1]?

After thinking about `DataOutput#canEmitBatchOfRecords`, I think that it looks strange to have `DataOutput` determine whether `it can continue to emit records`. The source of truth of `canEmitBatchOfRecords()` comes from things like `mailbox`, and it is more intuitive to keep it outside/decoupled from DataOutput.

[1] https://github.com/1996fanrui/flink/commits/30623/canEmitBatchOfRecords-POC

**Comment by 1996fanrui** [comment]
Hi @pnowojski @lindong28 , thanks for your discussion.

> About the Suppliers, I don't like using them personally tbh. For one thing they are hard to follow in the IDE (finding implementations or usages do not work). This could be mitigated with using a named interface instead of Supplier<Boolean>.
> 
> How about we add an internal functional interface in StreamTask? It could be similar to the existing SizeSupplier interface ([link](https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/metrics/groups/TaskIOMetricGroup.java#L294)).

I totally agree using a `named interface` instead of `Supplier<Boolean>`, and the `SizeSupplier` is a demo. How about this?

```
    /** Check whether records can be emitted in batch. */
    @FunctionalInterface
    public interface CanEmitBatchOfRecordsChecker {

        boolean check();
    }
```

> The other thing that looks strange if I take a step back, is that we now have to different while (true) loops. One in the MailboxProcessor#runMailboxLoop the other in SourceOperator and with [FLINK-30709](https://issues.apache.org/jira/browse/FLINK-30709) in the network source as well, doing basically the same thing.

Adding loops inside `SourceOperator` and `NetworkInput` is to reduce the call stack and avoid outer loops. The outer loop is still needed because `canEmitBatchOfRecords` may be false.

If we are worried about other developers misunderstanding, we can add some comments to the loop of `SourceOperator` and `NetworkInput`. What do you think?

**Comment by lindong28** [comment]
@1996fanrui The suggested approach sounds good to me.

**Comment by pnowojski** [comment]
> The other thing that looks strange if I take a step back, is that we now have to different while (true) loops. One in the MailboxProcessor#runMailboxLoop the other in SourceOperator and with [FLINK-30709](https://issues.apache.org/jira/browse/FLINK-30709) in the network source as well, doing basically the same thing.

I have thought about this a bit more, and I don't see an easy to implement alternative, so let's keep it as it is.

> After thinking about DataOutput#canEmitBatchOfRecords, I think that it looks strange to have DataOutput determine whether it can continue to emit records. The source of truth of canEmitBatchOfRecords() comes from things like mailbox, and it is more intuitive to keep it outside/decoupled from DataOutput.

For me it's the opposite thing :) Take a look at the `PushingAsyncDataInput`, that defines the internal APIs for pushing/pulling/emitting records. It would be strange for me that in order to fully implement this interface and use it as it's designed, one would have to ask "mailbox" or anything else. Why does the implementations (source operator/network input) should care where does this information come from? IMO Implementations should just emit record and check if they can immediately emit next one, ideally from the returned value of the `DataOutput#emitRecord` method. This pattern would be also consistent with returned status from the `PushingAsyncDataInput#emitNext` method. And on top of that, it would save us one extra method call, so should be slightly faster :)

**Comment by 1996fanrui** [comment]
I saw the `PushingAsyncDataInput#emitNext` and `SourceReader#pollNext` returns the `InputStatus`. It's similar to `DataOutput#emitRecord`. So sounds good to me, +1 for `DataOutput#emitRecord should return a boolean` determine whether `it can continue to emit records`. @lindong28 , what do you think?

If it's ok, `ReaderOutput#collect` and `RecordEmitter#emitRecord` should return a boolean as well. And the loop will be moved from `SourceOperator#emitNext` to `SourceReader#pollNext`. Because:

- `SourceReader#pollNext` just returns `InputStatus`, so `SourceOperator#emitNext` doesn't know whether it can continue to emit records.
-  All sub-class of `SourceReader` can do the loop inside of the `SourceReader#pollNext`. If it's right, we should do it in the `IteratorSourceReader` and `SourceReaderBase`.

Please correct me if I'm wrong, thanks~

**Comment by lindong28** [comment]
Hmm... if we add the method `boolean DataOutput#emitRecord(...)`, can you explain what Java Doc we should use to explain the semantics of the returned boolean value?

Intuitively, the semantics should describe an attribute/state of the DataOutput instance itself and should not change with the caller of this `DataOutput#emitRecord()`. This is the pattern followed by the existing `InputStatus` and `DataInputStatus` enum. However, it seems that in this case the returned boolean will change depending on whether it is called by `OneInputStreamTask` or `MultipleInputStreamTask`, which seems a bit counter intuitive. I am not sure if we can find a reasonable and intuitive semantics for this API.

**Comment by lindong28** [comment]
Thanks for the detailed explanation @pnowojski :) Let me provides comments below.

> It would be strange for me that in order to fully implement this interface and use it as it's designed, one would have to ask 
> "mailbox" or anything else.

With the approach using `CanEmitBatchOfRecordsChecker`, the implementation of the `DataOutput` interface does not rely on `mailbox`. Yes the caller would have to call `CanEmitBatchOfRecordsChecker#check` to use it as designated. 

With the approach using `boolean DataOutput#emitRecord`, now the implementation would have to reply on `mailbox` etc. And the caller needs to check the returned boolean and understand its semantics (the complexity depends on how we define its semantics). Also note that `SourceOperator#emitNext` is still checking things like `shouldWaitForAlignment` at every step in order to use DataOutput as designated. So I am not sure this approach is any less strange..

> This pattern would be also consistent with returned status from the PushingAsyncDataInput#emitNext method

I guess that probably depends on how we explain the semantics of the returned boolean. Maybe we can discuss more after seeing the proposed Java doc for this method.


> And on top of that, it would save us one extra method call, so should be slightly faster.

Hmm... I am not sure this approach can be faster. It seems that the same number of calls are made on the per-record runtime path. Though the SourceOperator will no longer call that supplier, the `DataOutput#emitRecord()` would now additionally call that supplier, right?


BTW, I do agree that implementations should just emit record and check if they can immediately emit next one. That is the key motivation of [FLINK-30533](https://issues.apache.org/jira/browse/FLINK-30533). But it seems that stream processing needs to have a lot of control flow work (e.g. new source split added/removed) and has to periodically check control flow messages. Batch processing does not have this issue. I also plan to see if there is to further mitigate this overhead in the future.

**Comment by pnowojski** [comment]
> If it's ok, ReaderOutput#collect and RecordEmitter#emitRecord should return a boolean as well. And the loop will be moved from SourceOperator#emitNext to SourceReader#pollNext.

I wouldn't go that far, at least not if strictly necessary. `SourceReader` is already part of the stable public API (`@Public`). Even if we could modify it without braking compatibility, it would make our future live more difficult, since we would have more complicated public API to support. And of course it would make the live of users implementing their sources more difficult as well.

> can you explain what Java Doc we should use to explain the semantics of the returned boolean value?

This should be simple:
```
Returns true if caller can call this method more times within a single invocation of the `PushingAsyncDataInput#emitNext` method. Otherwise returned value false means the caller should return from the `PushingAsyncDataInput#emitNext` invocation as soon as possible.
```
We do not need to explain anything about the multiple inputs, mailbox, record writer/state backend availability, etc. 

>  Also note that SourceOperator#emitNext is still checking things like shouldWaitForAlignment at every step in order to use DataOutput as designated

But waiting for alignment is a pure `SourceOperator`'s implementation detail. It has nothing to do with the output, or runtime availability.

> Batch processing does not have this issue. I also plan to see if there is to further mitigate this overhead in the future.

No, it doesn't. But batch doesn't care about single record latency and doesn't have checkpointing. Those two requirements combined, are The Cause why non blocking input/outputs/statebackend and mailbox exists in the first place. Without those requirements indeed I would also implement the code very differently. Just have a loop in the source that emits either a large batch of records or all records until it's finished, without any checks in the middle.

But I agree something doesn't feel right and I have a feeling that the overall design of the control flow could be improved. 

Btw, +1 @1996fanrui  for including the check of statebackend availability in this bug fix.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java [hunk 1]
```diff
@@ -993,7 +997,7 @@ public MailboxExecutorFactory getMailboxExecutorFactory() {
     }
 
     public boolean canEmitBatchOfRecords() {
-        return !this.mailboxProcessor.hasMail();
+        return !this.mailboxProcessor.hasMail() && taskIsAvailable();
     }
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:1001]: Could we have a unit test, that would check if this method returns a correct value? Ideally it would be great to have the assertion not only on the `StreamTask#canEmitBatchOfRecords()` level, but on the `DataOutput#...` level - to check that the value is correctly forwarded to the `SourceOperator` and network input.

**Comment by 1996fanrui** [comment]
>> If it's ok, ReaderOutput#collect and RecordEmitter#emitRecord should return a boolean as well. And the loop will be moved from SourceOperator#emitNext to SourceReader#pollNext.
>
> I wouldn't go that far, at least not if strictly necessary. SourceReader is already part of the stable public API (@Public). Even if we could modify it without braking compatibility, it would make our future live more difficult, since we would have more complicated public API to support. And of course it would make the live of users implementing their sources more difficult as well.

Sorry, I didn't realize that `ReaderOutput` and `SourceOutput` are also Public. 

The call chain of `NumberSequenceSource` is `SourceOperator#emitNext` -> `IteratorSourceReaderBase#pollNext` -> `SourceOutputWithWatermarks#collect` -> `DataOutput#emitRecord`.

The `SourceOutputWithWatermarks` and `ReaderOutput` implement `SourceOutput`, if `DataOutput#emitRecord should return a boolean` and we add the loop in the `SourceOperator` or `SourceReaderBase`,  we need add the return boolean for `SourceOutput#collect`. We will modify the `Public` api.

It should be compatible for user due to these `collect` methods don't have the return value, so users didn't use it before. We just add the return value. 

However, I'm not sure if it's a good idea, if `SourceOutput#collect` wants to return other state in the future, what should we do?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java (hunk 1 above)
Comment (1996fanrui) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:1001]: Sure, I can update all code after we discuss.

**Comment by lindong28** [comment]
```
Returns true if caller can call this method more times within a single invocation of the `PushingAsyncDataInput#emitNext` method. Otherwise returned value false means the caller should return from the `PushingAsyncDataInput#emitNext` invocation as soon as possible.
```

@pnowojski It seems that the API with this semantics is not self-contained. There is no clean way for the implementation of DataOutput#emitRecord() itself to enforce this semantic. And it kinds of couple the constructor of DataOutput with its user to be consistent. For example, a DataOutput created for `MultipleInputStreamTask` will always return false for this method, and thus can not be used by e.g. `OneInputStreamTask`. 

Ideally, we should allow the interface API to be self-contained so that it de-couples the implementation code from the caller code. What do you think?

**Comment by 1996fanrui** [comment]
Hi @lindong28 , as I understand, Piotr means `StreamTask` determines the `return value` of `boolean DataOutput#emitRecord`, but `boolean DataOutput#emitRecord` is self-contained for its caller.

`SourceOperator` and `NetworkInput` are caller of `boolean DataOutput#emitRecord`, they shouldn't care the logic of `boolean DataOutput#emitRecord`, they just use the `return value` to do some corresponding logic. So `boolean DataOutput#emitRecord` is self-contained for its caller.

In other words: A and B call the interface of C, and the `return value` of the interface of C is determined by D. I think the interface of C should be self-contained for A and B. And A and B should not care about the interface of C is determined by D. This should be a common scenario, that is: the logic or `return value` of the interface is affected by some parameters.

And please correct me if I'm wrong. 

cc @pnowojski

**Comment by pnowojski** [comment]
Yes, exactly @1996fanrui.

@lindong28, indeed, there is an issue that incorrectly implemented caller can ignore the return value. But how is it worse compared to your proposal of passing separate `Supplier<Boolean>`? I would say that your proposal is even more risky, since the `Supplier<Boolean>` would be independent of the `DataOutput`, so it's even more likely that someone implementing/using `DataOutput` would forget or even not realise that there is some `Supplier<Boolean>` that should be passed and used per every `void DataOutput#emitRecord` call.

**Comment by lindong28** [comment]
@1996fanrui @pnowojski Regarding whether `boolean DataOutput#emitRecord`, maybe I can use the example provided by Rui above to elaborate my concern. 

Let's say we have `MultiInputStreamTask-D` and `OneInputStreamTask-B`. And `MultiInputStreamTask-D` constructed a `DataOutput-C` which is somehow used by `OneInputStreamTask-B`.

Now we know that `DataOutput-C::emitRecord` should always return false since it is constructed for `MultiInputStreamTask-D`. This means when `OneInputStreamTask-B` calls `DataOutput-C::emitRecord`, it will always get `false`, which effectively disables the optimization we are trying to make. This is because `DataOutput-C::emitRecord`  fails to enforce its semantics when it is used by `OneInputStreamTask-B`.

Do you think this example makes sense? Can you help me understand which part of this explanation is incorrect? Thanks.

**Comment by 1996fanrui** [comment]
Hi @lindong28 
It's strange that `MultiInputStreamTask-D` constructed a `DataOutput-C`, and  `OneInputStreamTask-B` calls the `DataOutput-C::emitRecord` constructed for `MultiInputStreamTask-D`. I think it should be a bug if it happens.

`DataOutput::emitRecord` semantics should be correct if each `StreamTask` calls the corresponding `DataOutput`. If the correct DataOutput is not guaranteed to be called, there will be many problems with the code.

For example: `MultipleInputStreamTask`[1] and `OneInputStreamTask`[2] have corresponding `CheckpointBarrierHandler`, if `MultipleInputStreamTask` calls `CheckpointBarrierHandler` of `OneInputStreamTask`, there should be similar bugs.

`CheckpointBarrierHandler` is similar. `MultipleInputStreamTask` and `OneInputStreamTask` create different `CheckpointBarrierHandlers`, and then call `checkpointBarrierHandler.processBarrier`. `StreamTask` does not need to care about the logic inside `checkpointBarrierHandler.processBarrier`. If there is a `return value`, the caller are only responsible for using the `return value`, and don't care about how the `return value` is generated. Similarly, there may be different return values depending on the StreamTask.

[1] https://github.com/apache/flink/blob/0b8a83ce54d39d0d5a5b82573c5037f306e9f7f7/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/MultipleInputStreamTask.java#L73
[2] https://github.com/apache/flink/blob/0b8a83ce54d39d0d5a5b82573c5037f306e9f7f7/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTask.java#L67

**Comment by lindong28** [comment]
> It's strange that MultiInputStreamTask-D constructed a DataOutput-C, and OneInputStreamTask-B calls the DataOutput-
> C::emitRecord constructed for MultiInputStreamTask-D. I think it should be a bug if it happens.

@1996fanrui Hmm... I thought we have agreed that "A and B should not care about the interface of C is determined by D". This is because DataOutput is an interface and the caller should be able to use it based on its API semantics without knowing who constructed it, right?

Do you suggest that we should now couple the caller the DataOutput with the code that constructed it? It seems to break the typical pattern of how we use interface..

**Comment by lindong28** [comment]
> For example: MultipleInputStreamTask[1] and OneInputStreamTask[2] have corresponding CheckpointBarrierHandler, if MultipleInputStreamTask calls CheckpointBarrierHandler of OneInputStreamTask, there should be similar bugs.

Indeed, ultimately the caller will have to choose the right caller instance. On the other hand, the interface or abstract method is typically expected to have clean and self-contained semantics so that it de-couples the caller from the method implementation as much as possible.

`boolean DataOutput#emitRecord` indeed would work as long as it is always called by the same StreamTask instance that constructed it. But it adds more coupling here and you would have to explain this API's behavior based on who is the caller and who constructed this DataOutput, which IMO is an anti-pattern for abstract/interface methods.

If you both think that this approach is cleaner and it is OK to have an abstract method's semantics depend on its caller, I am fine to stop here and let this PR in.

**Comment by 1996fanrui** [comment]
> Are you suggesting that we should now couple the caller of DataOutput with the code that constructed it?

In fact, the code in the master branch did this a long time ago. 

`OneInputStreamTask` has inner class `StreamTaskNetworkOutput`, and `OneInputStreamTask#createDataOutput` will create it. 

`StreamMultipleInputProcessorFactory` has inner class `StreamTaskSourceOutput`, and `MultipleInputStreamTask#createInputProcessor -> StreamMultipleInputProcessorFactory#create` will create it.

**Comment by pnowojski** [comment]
@lindong28 I don't understand your concerns. With separate functional interface as you are proposing you will have exact same "problems". If you add functional interface to the [AbstractStreamTaskNetworkInput](https://github.com/apache/flink/pull/21697/commits/328510c28779d64fa249c33f6a1014fb7c7f6ddf#diff-68389ccd4de1418caa25baa2429ebdc1040834f5386ecb046acc940c418086ba), you would have exact same issue, that you would need to pass different instances of that interface, that would at the moment return always `false` for multiple input tasks (note IMO that's only a temporary limitation. `MultipleInputStreamTask` could be implemented in the future to allow batching).

On top of that, I really don't see this being an issue at all. That's the whole point of code re-usability, regardless if it's object oriented or functional. You want to pass different implementations, that will be behaving differently.

> Are you suggesting that we should now couple the caller of DataOutput with the code that constructed it?

Caller shouldn't care if it's coupled or not. It should just adhere to the contracts of the instances/classes/functions that are passed to it. And in this case the caller indeed doesn't care at all `SourceOperator` instances are using the same `DataOutput` instance or not. Just use the instance that's passed.

For me the only difference between (1) separate functional interface vs embedding the same functionality in (2) the `DataOutput` is that the first one behaviour how the `DataOutput` should be used is controlled and passed independently of it, which hides the connection and increases chances that someone will forget to use the functional interface after some refactor/adding new code. And it's more verbose as you need to pass around two objects instead of just one.

**Comment by 1996fanrui** [comment]
Hi @pnowojski , after an offline discussion with @lindong28 , I'm not sure if it makes sense to `add a return value for DataOutput#emitRecord` and `the return value determines whether the caller can continue to emit the record`?

`DataOutput#emitRecord` is an interface, it should obey these conditions:
- It shouldn't care about the logic of the caller.
- It can be called repeatedly by the caller (Most methods can be called repeatedly by the caller, except for some special methods, such as: `open()`, `close()`, etc.).
- It shouldn't care what the caller does between calling itself multiple times.

Simplified code for our discussion looks like this:

```java
MailboxProcessor#runMailboxLoop() {
  while(true) {
    // do some logics

    NetworkInput/SourceOperator#emitNext();

    // do some logics
  }
}

// Before change: just emit one record.
NetworkInput/SourceOperator#emitNext(){
  DataOutput#emitRecord();
}

// After change: can emit multiple records.
NetworkInput/SourceOperator#emitNext(){
  while(true) {
    DataOutput#emitRecord();
  }
}
```

We want to add the loop inside `emitNext` to reduce the call stack and improve performance. The `loop inside emitNext` logic should belong to `NetworkInput/SourceOperator` instead of `DataOutput`.

And I mentioned before: `DataOutput#emitRecord can be called repeatedly by the caller`. If we return a boolean to determine whether caller can continue to emit records, it means that `DataOutput#emitRecord` knows the caller's two loops, however, `it shouldn't care about the logic of the caller`, and it shouldn't know the logic about `canEmitBatchOfRecords`. It should only be responsible for emitting records.

So I think the logic related to the `canEmitBatchOfRecords` should be maintained inside the `NetworkInput and SourceOperator`, and I have a POC branch to do it[1].

Also, since flink 1.17 is approaching the feature freeze, could I create a new PR to fix bug and this PR focus on the refactoring? We can continue discuss the refactoring here, what do you think?

[1] https://github.com/1996fanrui/flink/commits/30623/canEmitBatchOfRecords-POC

**Comment by pnowojski** [comment]
Hey, again I think your arguments are not fully correct:

> It can be called repeatedly by the caller (Most methods can be called repeatedly by the caller, except for some special methods, such as: open(), close(), etc.).
> It shouldn't care what the caller does between calling itself multiple times.

Both are not true. There are multiple places where methods number of calls and order of calls matter a lot. The most similar are our non blocking outputs. Callers of `RecordWriter.emit(...)` should check for the output availability, for example. The same applies to availability of the statebackends (currently only changelog). 

Also keep in mind, that braking out of the inner loop, is not an optimisation, **but a hard requirement**. Before FLINK-30533  `DataOutput#emitNext` had actually an implicit contract, forbidding the callers (`SourceOperator` or `NetworkInput`) to loop over `DataOutput#emitNext`, as otherwise that would cause starvation, braking checkpointing mechanism among many other issues. Even with your current implementation, multiple/two input  are still forbidden to loop.

> It shouldn't care about the logic of the caller.

So that's not true as well. Regardless how you implement it, the caller has to care about how many times the `DataOutput#emitRecord()` can be called in a row. If you hide this logic in a method/interface that passed separately, it doesn't make it disconnected in any way. Instead of passing one interface with both `emitRecord()` and something that returns `canEmitBatchOfRecords()` (either via return value or separate method in the `DataOutput` interface), you have to pass two interfaces. One that controls how the other should be used. I don't see any benefit of this approach:
1. caller still needs both of the the things passed
2. even for testing and dependency injection, you still need to implement the same mock logic, having them separate doesn't simplify anything here

And as I mentioned before, passing two separate interface would make bugs much easier to appear, when someone forgets about `CanEmitBatchOfRecordsChecker`.

**Comment by lindong28** [comment]
> So that's not true as well. Regardless how you implement it, the caller has to care about how many times the DataOutput#emitRecord() can be called in a row. If you hide this logic in a method/interface that passed separately, it doesn't make it disconnected in any way. Instead of passing one interface with both emitRecord() and something that returns canEmitBatchOfRecords() (either via return value or separate method in the DataOutput interface), you have to pass two interfaces. One that controls how the other should be used. I don't see any benefit of this approach:

@pnowojski It seems that we are talking about different things here. I think Rui is saying that "callee should not care about the caller". And you are saying that "caller has to care about how many times callee can be called". IMO both statements are correct.

This is the main reason I am against the method `boolean DataOutput#emitRecord` -- it breaks the commonly accepted principle that "the definition of interface method should be self-descriptive and not depend on what caller does". If an interface method does not follow this principle, IMO it pretty much degrades to a class as it introduces bi-directional ties between caller and callee.

I also asked @becketqin offline to comment on this point but I am not sure if he has time.

Anyway, in case we can not agree on this point anytime soon, do you mind if we create a new PR to fix the bug and let this PR focus on the refactoring?

**Comment by pnowojski** [comment]
Ok, let's go with the `CanEmitBatchOfRecordsChecker`

**Comment by MartijnVisser** [comment]
Just to double check, is this PR nearing a stage where it can be merged soon? Especially since this performance regression has been open for quite some time and is a blocker for the 1.17 release

**Comment by 1996fanrui** [comment]
> Ok, let's go with the `CanEmitBatchOfRecordsChecker`

Hi @pnowojski , thanks for your feedback, I will update the PR today.



> Just to double check, is this PR nearing a stage where it can be merged soon? Especially since this performance regression has been open for quite some time and is a blocker for the 1.17 release

Hi @MartijnVisser , thanks for your check. I think this PR can be merged this week or early next week.

**Comment by MartijnVisser** [comment]
> I think this PR can be merged this week or early next week.

Thanks. We have to make sure that this is merged before the 31st, since that's when the branch cut will happen.

**Comment by 1996fanrui** [comment]
Hi @pnowojski @lindong28 , I have updated the PR and added some tests to check the result of `CanEmitBatchOfRecordsChecker`, please help take a look in your free time, thanks~

**Comment by 1996fanrui** [comment]
@flinkbot run azure

**Comment by 1996fanrui** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
@1996fanrui It looks like you need to push an empty commit or rebase/force push again, CiBot doesn't seem to pick up your latest commits but I see it work for other PRs

**Review by pnowojski — APPROVED** [review]
Thanks for the update. LGTM assuming azure is green

**Review by lindong28 — APPROVED** [review]
Thanks @pnowojski @1996fanrui for the discussion and the PR! LGTM.

**Comment by 1996fanrui** [comment]
There are some unit tests fail due to the first commit, and I have updated.

The `Supplier<Boolean> mailboxHasMail;` should be true in the `TestingSourceOperator` by default, however the `canEmitBatchOfRecords` should be false by default.

**Comment by 1996fanrui** [comment]
@pnowojski @lindong28  Thanks a lot for the detailed discussion and suggestion.
